### DISCUSSION TITLE
It might not be possible to reference types in other types

### DISCUSSION THREAD

**Discussion by MichaelXavier** [chrisdone/hell#64] [Q&A] [answered]
Given the minimal example file:

```
data Sum = Foo | Bar

data Baz = Baz
  { sum :: Sum
  }

main = do
  Text.putStrLn "hello world"
```

You'll get an error along the lines of `Unknown type: TyCon "Sum"`. You get a similar thing if you reference Sum by `Main.Sum`. If you instead define some other record `Quux` and add a field to `Baz` of that type, you get the same kind of error. Because this affects both records (desugars to a Record type) and sums (desugars to a Variant type), maybe the syntactic sugar allows you to handle the values directly but isn't available to the type system?

**Comment by chrisdone** [comment]
That's right. Presently you can't reference user-defined types in your user-defined types. I plan to lift that limitation. :ok_hand: Just needs a simple dependency analysis and then referencing `Sum` in `Baz` would be like a type-alias expanded at desugaring time to the tagged variant `Tagged "Sum" ...`.

But recursively defined types present a challenge, so those would in the near-term not be supported.

**Reply by chrisdone** [reply]
It's probably rare that someone would want to define a recursive type for a shell script, so not sure it's worth the effort. There's also `Data.Tree` that could serve a lot of cases for scripting, which I was thinking to add to support the "tree-like structures" use-case.

**Comment by MichaelXavier** [comment]
@chrisdone Just saw that you added referencing custom types within records and snatched it up into my config to test it out. Nesting record types within records seems to work as expected, but sums seem to have a quirk. If you do this:

```haskell
data MyRecord = MyRecord {sum :: Main.MySum}

data MySum = MySumL | MySumR

main = do
  let myRecord = Main.MyRecord {sum = Main.MySumR}
  Text.putStrLn "hello world"
```

You'll get an error:
```
hell: Unification error: Couldn't match type

  "Main.MySum"

against type

  "MySum"
```

I tried a few different permutations but it seems like the sum must be constructed with the `Main` module qualifier.

**Reply by chrisdone** [reply]
I fixed this in [Ensure user-defined sum/record type are qualified by chrisdone](https://github.com/chrisdone/hell/pull/68). It was as suspected, lack of qualification. I should use a more disciplined type for names, but was thus far getting away with dodging it. ;-)

**Comment by chrisdone** [comment, marked as answer]
Oh, thanks for trying it! Looks like I’ve been inconsistent in a key area in terms of generating the name of the type. It should be qualified, and probably it was not qualified previously in the constructor.

I’ll fix this when I’m back at keyboard. 

On Wed, Jan 1, 2025, at 8:06 PM, Michael Xavier wrote:
> 
> 
> @chrisdone <https://github.com/chrisdone> Just saw that you added referencing custom types within records and snatched it up into my config to test it out. Nesting record types within records seems to work as expected, but sums seem to have a quirk. If you do this:
> 
> data MyRecord = MyRecord {sum :: Main.MySum}
> 
> data MySum = MySumL | MySumR
> 
> main = do
>   let myRecord = Main.MyRecord {sum = Main.MySumR}
>   Text.putStrLn "hello world"
> You'll get an error:
> 
> `hell: Unification error: Couldn't match type
> 
>   "Main.MySum"
> 
> against type
> 
>   "MySum"
`

**Reply by MichaelXavier** [reply]
That fixed it! Thank you!
